internal/mage/util/engine.go

```diff
@@ -344,65 +345,136 @@ func devEngineContainersWithGPUSupport(ctx context.Context, c *dagger.Client, ar
-		Export(ctx, tarballPath)
-	if err != nil {
-		panic(err)
+func typescriptSDKContent(ctx context.Context, c *dagger.Client, arch string) dagger.WithContainerFunc {
+	return func(ctr *dagger.Container) *dagger.Container {
+		sdkCtrTarball := c.Container().
+			WithRootfs(c.Host().Directory("sdk/typescript", dagger.HostDirectoryOpts{
+				Include: []string{
+					"**/*.ts",
+					"LICENSE",
+					"README.md",
+					"runtime",
+					"package.json",
+					"dagger.json",
+				},
+				Exclude: []string{
+					"node_modules",
+					"dist",
+					"**/test",
+					"**/*.spec.ts",
+					"dev",
+				},
+			})).
+			WithFile("/codegen", goSDKCodegenBin(c, arch)).
+			AsTarball(dagger.ContainerAsTarballOpts{
+				ForcedCompression: dagger.Uncompressed,
+			})
+
+		sdkDir := c.Container().From("alpine:"+alpineVersion).
+			WithMountedDirectory("/out", c.Directory()).
+			WithMountedFile("/sdk.tar", sdkCtrTarball).
+			WithExec([]string{"tar", "xf", "/sdk.tar", "-C", "/out"}).
+			Directory("/out")
+
+		var index ocispecs.Index
+		indexContents, err := sdkDir.File("index.json").Contents(ctx)
+		if err != nil {
+			panic(err)
+		}
+		if err := json.Unmarshal([]byte(indexContents), &index); err != nil {
+			panic(err)
+		}
+		manifest := index.Manifests[0]
+		manifestDgst := manifest.Digest.String()
+
+		return ctr.
+			WithEnvVariable(distconsts.TypescriptSDKManifestDigestEnvName, manifestDgst).
+			WithDirectory(distconsts.EngineContainerBuiltinContentDir, sdkDir, dagger.ContainerWithDirectoryOpts{
+				Include: []string{"blobs/"},
+			})
 	}
+}
 
-	f, err := c.Host().File(tarballPath).Sync(ctx)
-	if err != nil {
-		panic(err)
+func goSDKContent(ctx context.Context, c *dagger.Client, arch string) dagger.WithContainerFunc {
+	return func(ctr *dagger.Container) *dagger.Container {
+		base := c.Container(dagger.ContainerOpts{Platform: dagger.Platform("linux/" + arch)}).
+			From(fmt.Sprintf("golang:%s-alpine%s", golangVersion, alpineVersion))
+
+		sdkCtrTarball := base.
+			WithEnvVariable("GOTOOLCHAIN", "auto").
+			WithFile("/usr/local/bin/codegen", goSDKCodegenBin(c, arch)).
+			WithEntrypoint([]string{"/usr/local/bin/codegen"}).
+			AsTarball(dagger.ContainerAsTarballOpts{
+				ForcedCompression: dagger.Uncompressed,
+			})
+
+		sdkDir := base.
+			WithMountedDirectory("/out", c.Directory()).
+			WithMountedFile("/sdk.tar", sdkCtrTarball).
+			WithExec([]string{"tar", "xf", "/sdk.tar", "-C", "/out"}).
+			Directory("/out")
+
+		var index ocispecs.Index
+		indexContents, err := sdkDir.File("index.json").Contents(ctx)
+		if err != nil {
+			panic(err)
+		}
+		if err := json.Unmarshal([]byte(indexContents), &index); err != nil {
+			panic(err)
+		}
+		manifest := index.Manifests[0]
+		manifestDgst := manifest.Digest.String()
+
+		return ctr.
+			WithEnvVariable(distconsts.GoSDKManifestDigestEnvName, manifestDgst).
+			WithDirectory(distconsts.EngineContainerBuiltinContentDir, sdkDir, dagger.ContainerWithDirectoryOpts{
+				Include: []string{"blobs/"},
```

Comment:
Neat - I've never actually realized that `WithDirectory` copies additively to the target directory. :exploding_head: